cli/npm/installer/resolution.rs

```diff
@@ -257,16 +259,66 @@ fn populate_lockfile_from_snapshot(
     let dependencies = pkg
       .dependencies
       .iter()
-      .map(|(name, id)| NpmPackageDependencyLockfileInfo {
-        name: name.clone(),
-        id: id.as_serialized(),
+      .filter_map(|(name, id)| {
+        if pkg.optional_dependencies.contains(name) {
+          None
+        } else {
+          Some(NpmPackageDependencyLockfileInfo {
+            name: name.clone(),
+            id: id.as_serialized(),
+          })
+        }
+      })
+      .collect();
+
+    let optional_dependencies = pkg
+      .optional_dependencies
+      .iter()
+      .filter_map(|name| {
+        let id = pkg.dependencies.get(name)?;
+        Some(NpmPackageDependencyLockfileInfo {
+          name: name.clone(),
+          id: id.as_serialized(),
+        })
       })
       .collect();
 
+    let optional_peers = pkg
+      .optional_peer_dependencies
+      .iter()
+      .filter_map(|name| {
+        let id = pkg.dependencies.get(name)?;
+        Some(NpmPackageDependencyLockfileInfo {
+          name: name.clone(),
+          id: id.as_serialized(),
+        })
+      })
```

Comment:
need to clarify if this is included in `pkg.dependencies` and if `optionalPeers` should be only unresolved ones